clang/lib/Parse/ParseDecl.cpp

```diff
@@ -2554,16 +2554,27 @@ Decl *Parser::ParseDeclarationAfterDeclarator(
   return ParseDeclarationAfterDeclaratorAndAttributes(D, TemplateInfo);
 }
 
+static bool isConstexprVariable(const Decl *D) {
+  if (const VarDecl *Var = dyn_cast_if_present<VarDecl>(D))
+    return Var->isConstexpr();
+
+  return false;
+}
+
 Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
     Declarator &D, const ParsedTemplateInfo &TemplateInfo, ForRangeInit *FRI) {
   // RAII type used to track whether we're inside an initializer.
   struct InitializerScopeRAII {
     Parser &P;
     Declarator &D;
     Decl *ThisDecl;
+    EnterExpressionEvaluationContext EC;
 
     InitializerScopeRAII(Parser &P, Declarator &D, Decl *ThisDecl)
-        : P(P), D(D), ThisDecl(ThisDecl) {
+        : P(P), D(D), ThisDecl(ThisDecl),
+          EC(P.Actions, Sema::ExpressionEvaluationContext::ConstantEvaluated,
+             ThisDecl, Sema::ExpressionEvaluationContextRecord::EK_Other,
+             isConstexprVariable(ThisDecl)) {
```

Comment:
This doesn't look right.  The relevant standard text says:

- "An expression or conversion is in an immediate function context if it is potentially evaluated and [...] it is a subexpression of a manifestly constant-evaluated expression or conversion [...]"
- "An expression or conversion is manifestly constant-evaluated if it is [...] the initializer of a variable that is usable in constant expressions or has constant initialization"

So we need to handle:
- constexpr variables
- variables of type `const int`
- constant-initialized variables (global/static)

Reply:
On a side-note, gcc currently accepts the following; if I'm understanding the rule, it's not supposed to be valid.

```
struct S {
  int *value;
  constexpr S(int v) : value(new int {v}) {}
  constexpr ~S() { delete value; }
};
consteval S fn() { return S(5); }
int g();
void f() {
  int a = *fn().value;
}
```

Reply:
> if I'm understanding the rule, it's not supposed to be valid.

Agreed; MSVC appears to get this right:
```
<source>(8): error C7595: 'fn': call to immediate function is not a constant expression
```

Since `int a` is not usable in a constant expression, its initializer isn't manifestly constant-evaluated. Thus only the call to `fn()` is manifestly constant-evaluated, so the access to the heap-allocated memory is ill-formed.

...as far as I can tell 😅

Reply:
@efriedma-quic I see your point - we need to handle a broader class of variables in this case than just `constexpr`. Let me see what I can do, thanks for the feedback.

Reply:
@efriedma-quic Your feedback led to my taking a different approach for this PR; please see [this comment](https://github.com/llvm/llvm-project/pull/89565#issuecomment-2079917447).